File: combat/DamageCalculator.cs
using System;
using Lanes.Combat.Units;

namespace Lanes.Combat;

/// <summary>
/// Stateless utility that resolves the numerical outcome of an ability use.
/// All formulae live here so they can be tweaked, tested, and balanced in
/// isolation from the rest of the combat system.
///
/// Formula (physical):
///   raw    = attacker.Attack * ability.PowerMultiplier
///   damage = max(1, raw - target.Defense)
///
/// Formula (magical):
///   raw    = attacker.MagicAttack * ability.PowerMultiplier
///   damage = max(1, raw - target.MagicDefense)
///
/// Formula (healing):
///   amount = attacker.MagicAttack * ability.PowerMultiplier  (min 1)
///   Healing bypasses defence; it restores HP instead.
/// </summary>
public static class DamageCalculator
{
    /// <summary>
    /// Calculates the amount of damage (or healing) produced by
    /// <paramref name="attacker"/> using <paramref name="ability"/>.
    /// The sign is always positive; call sites decide whether to subtract
    /// (damage) or add (healing) to the target's HP.
    /// </summary>
    public static int Calculate(UnitInstance attacker, AbilityDefinition ability)
    {
        if (attacker is null) throw new ArgumentNullException(nameof(attacker));
        if (ability is null)  throw new ArgumentNullException(nameof(ability));

        float raw = ability.DamageType switch
        {
            DamageType.Physical => attacker.Stats.Attack * ability.PowerMultiplier,
            DamageType.Magical  => attacker.Stats.MagicAttack * ability.PowerMultiplier,
            DamageType.Healing  => attacker.Stats.MagicAttack * ability.PowerMultiplier,
            _ => throw new ArgumentOutOfRangeException(nameof(ability), "Unknown DamageType.")
        };

        return Math.Max(1, (int)raw);
    }

    /// <summary>
    /// Calculates the net damage dealt to <paramref name="target"/> after
    /// defence reduction.  Returns 0 for healing abilities (no defence check).
    /// </summary>
    public static int CalculateWithDefense(UnitInstance attacker, UnitInstance target, AbilityDefinition ability)
    {
        if (target is null) throw new ArgumentNullException(nameof(target));
        if (ability.DamageType == DamageType.Healing)
            return 0;

        int raw = Calculate(attacker, ability);
        int defense = ability.DamageType switch
        {
            DamageType.Physical => target.Stats.Defense,
            DamageType.Magical  => target.Stats.MagicDefense,
            _ => 0
        };

        return Math.Max(1, raw - defense);
    }
}

File: src/app/core/engine/damage-calculator.ts
import { AbilityDefinition } from '../models/ability.model';
import { UnitInstance } from '../models/unit-instance.model';

/**
 * Calculates raw and net combat numbers for abilities.
 */
export class DamageCalculator {
  public static hasAdvantage(attacker: UnitInstance, target: UnitInstance): boolean {
    if (attacker.team === target.team) return false;
    const from = attacker.classDef.archetype;
    const to = target.classDef.archetype;
    return (
      (from === 'melee' && to === 'ranged') ||
      (from === 'ranged' && to === 'magic') ||
      (from === 'magic' && to === 'melee')
    );
  }

  /**
   * Computes the raw power output before defence reduction.
   */
  public static calculateRaw(attacker: UnitInstance, ability: AbilityDefinition): number {
    let raw = 0;
    switch (ability.damageType) {
      case 'physical':
        raw = attacker.effectiveAttack * ability.powerMultiplier;
        break;
      case 'magical':
        raw = attacker.stats.magicAttack * ability.powerMultiplier;
        break;
      case 'healing':
        raw = attacker.stats.magicAttack * ability.powerMultiplier;
        break;
    }
    return Math.max(1, Math.floor(raw));
  }

  /**
   * Computes net damage after defense (or healing without defense check).
   */
  public static calculateWithDefense(
    attacker: UnitInstance,
    target: UnitInstance,
    ability: AbilityDefinition,
  ): number {
    if (ability.damageType === 'healing') {
      return this.calculateRaw(attacker, ability);
    }

    const raw = this.calculateRaw(attacker, ability);
    const defense =
      ability.damageType === 'physical' ? target.effectiveDefense : target.stats.magicDefense;

    const net = Math.max(1, raw - defense);
    // MVP tuning: the tactical triangle adds 25% after defense, with no reverse penalty.
    return Math.floor(net * (this.hasAdvantage(attacker, target) ? 1.25 : 1));
  }
}
